### DISCUSSION TITLE
Modifiied repo to use PID Climate

### DISCUSSION THREAD

**Discussion by Karl-opec** [nliaudat/floor-heating-controller#11] [Show and tell]
I have commited my modifications to this repo in [Karl-Opec](https://github.com/Karl-opec/floor-heating-controller)  to use Climate PID and bluetooth Temperature Sensor for the Rooms.

Any Comments welcome

**Comment by nliaudat** [comment]
Hi, 
Do you have good results with PID ?

How many times an hour your valves changes their status ?
_I had bad results with PID who was making too many valve position changes. 
Here you'll find mine for last 24h_

![image](https://github.com/nliaudat/floor-heating-controller/assets/6782613/1c5f979e-8667-4af9-bb83-31f3f52419c4)

I have pretty smooth temperature in living room according to outdoor temperature changes : 

![image](https://github.com/nliaudat/floor-heating-controller/assets/6782613/aecc9373-134d-4401-8905-97bfdb682ea9)

**Comment by Karl-opec** [comment]
See the following:

![grafik](https://github.com/nliaudat/floor-heating-controller/assets/1394350/1871a8e6-3caf-48a7-bb8b-9d047bf38a8e)

In the left side of the chart you see the stable Temp Soll/Ist  mean desired-/measured- Temp. Then a window is opened and the measured temp drops, the Öffnung (valve open State) goes up to 100%. Steps are min-movement% appx. every 3-4 minutes. The PID's get a temp value ervery 60sec (heart-beat) of sensor.  This means every 60sec PID loop is recalced.

It works fine

**Comment by nliaudat** [comment]
Glad it works well !

You mentionned github/Technologic in config.yaml to find the correct pid values. I cannot find that repo :(

```
# PID Parameter github/Technologic values kp=0.05 ki=0,0001 kd=0 ; aktuell update temp ist ca.2,5 min > cycle time 150 sec daher 0,05/(8h*3600s/150s)) 0,00026
  pid_kp: "0.06"
  pid_ki: "0.00011"
  pid_kd: "4.0"
```

How do you compute your pid values ? 
I've used the formula from https://www.mdpi.com/1996-1073/13/8/2068 without good results

I''ll make a try on your solution to compare but my floor heating is low temperature and has a lot of inertia. I mean that I cannot see major differences between pid and "differential target temp" systems. 

Can You ?

Regards

**Reply by nliaudat** [reply]
found repo sorry : https://github.com/Tecnologic/floor-heating-controller

**Comment by heinemannj** [comment]
I'm actually testing a return flow based thermostat.

1-wire temperature sensors for each heating circuit:
a) Supply flow
b) Return flow

Initially I've started with the "differential target temp" approach (return flow target) which was working more or less accurate - return flow based temperature measurements are extremely sensitive - you can measure temp diffs of ~0.1K within a minute!

Second I've tried your setup of PID => my ESP is rebooting latest after 16 mins ...

```
WARNING floor-heating-1: Connection error occurred: [Errno 104] Connection reset by peer
INFO Processing unexpected disconnect from ESPHome API for floor-heating-1
WARNING Disconnected from API
WARNING Can't connect to ESPHome API for floor-heating-1: Error connecting to ('192.168.x.x', 6053): [Errno 111] Connect call failed ('192.168.x.x', 6053) (SocketAPIError)
INFO Trying to connect to floor-heating-1 in the background
INFO Successfully connected to floor-heating-1 in 0.029s
INFO Successful handshake with floor-heating-1 in 0.172s
```

Switching back to the roots.

Any ideas:?

**Reply by nliaudat** [reply]
Perhaps your sensor must be less sensitive ...

https://esphome.io/components/climate/pid.html : 
_We recommend putting a filter on the sensor (see filters in [Sensor Component](https://esphome.io/components/sensor/)) and using output_averaging_samples to calm the PID sensor from a noisy input sensor._

**Comment by nliaudat** [comment]
> I have commited my modifications to this repo in [Karl-Opec](https://github.com/Karl-opec/floor-heating-controller) to use Climate PID and bluetooth Temperature Sensor for the Rooms.
> 
> Any Comments welcome

Perhaps it could be a good idea to make a separate branch for your Climate PID integration in that repo. What do you mean ?

**Reply by heinemannj** [reply]
I've made a lot of code changes:

- naming od IDs and names
- additional temp sensors
- different cron jobs
- adaptations based to my HA dashboard
- ...

Hard to sync.

I will publish my repro within the next days.

Then we can briefly discuss to develop a STANDARD.

**Comment by Karl-opec** [comment]
> I'm actually testing a return flow based thermostat.
> 
> 1-wire temperature sensors for each heating circuit: a) Supply flow b) Return flow
> 
> Initially I've started with the "differential target temp" approach (return flow target) which was working more or less accurate - return flow based temperature measurements are extremely sensitive - you can measure temp diffs of ~0.1K within a minute!
> 
> Second I've tried your setup of PID => my ESP is rebooting latest after 16 mins ...
> 
> ```
> WARNING floor-heating-1: Connection error occurred: [Errno 104] Connection reset by peer
> INFO Processing unexpected disconnect from ESPHome API for floor-heating-1
> WARNING Disconnected from API
> WARNING Can't connect to ESPHome API for floor-heating-1: Error connecting to ('192.168.x.x', 6053): [Errno 111] Connect call failed ('192.168.x.x', 6053) (SocketAPIError)
> INFO Trying to connect to floor-heating-1 in the background
> INFO Successfully connected to floor-heating-1 in 0.029s
> INFO Successful handshake with floor-heating-1 in 0.172s
> `
> 
> Switching back to the roots.
> 
> Any ideas:?

Rebooting the ESP is happening when timing problems happen. Try using ESP-IDF. I have disabled api to homeassistant. 
mine ESP-32 runs at least 4 to five days before it autostarts again.

I envision you use the dallas 18B20 sensors to mesure the  Supply and Return Flow temperatures? If so, these sensors are very calculation time resource vampire's. 

I use in a different situation in total 18 Dallas 18B20 with i2c DS2482-800 chip which communicates per i2c with the CPU and take a lot of computing burden away from thee ESP-32.

These are my quick thoughts

**Reply by nliaudat** [reply]
Have you tried to disable 1-wire and see if the reboots stops ? (1-wire is a complex protocol with a big library)

For sensors, I mean the best is [NTC sensors](https://esphome.io/components/sensor/ntc.html) : 

- It's cheap
- You can easyly fix it on your tubes
- There no risk of corrosion
- Analog

You could use an analog to digital [multiplexer](https://esphome.io/components/sensor/cd74hc4067.html) with only one ESP32 ADC pin

**Comment by heinemannj** [comment]
Many thanks to both of you for your input and exellent support!

I've done a little bit of code cleanup and increaseed the update time of a couple of sensors.
NOT touched api to homeassistant.
NOT touched my 1-wire sensors.

**Switched from Arduino framework to esp-idf**:

```
#esp32:
#  board: esp32dev
#  framework:
#    type: arduino
esp32:
  board: nodemcu-32s
  framework:
    type: esp-idf
    version: recommended
    sdkconfig_options:
      CONFIG_COMPILER_OPTIMIZATION_SIZE: y
```

**Since ~11 hours no reboot!**

Because I've made a couple of changes at the same time I can not be 100% sure, if the change to esp-idf has resolved - but on 98%.

Any risks or disadvantages by using esp-idf?

Default ESPHome settings after ESP32 device setup:
```
esp32:
  board: esp32dev
  framework:
  type: arduino
```

**Reply by nliaudat** [reply]
In the past, esp-idf framework was missing functionnalities like web interface and async http requests. 
Actually, I mean it perform better than arduino cause the memory management is very well done in esp-idf. 
I have migrated all my project to idf without troubles.

**Reply by heinemannj** [reply]
Can you please provide your PID parameter?
Have you already an up and running PID setup?

**Comment by nliaudat** [comment]
Hi, 
I've created a PID test in the [latest release (v108)](https://github.com/nliaudat/floor-heating-controller/commit/f9da2bcf62a5e6bed01520066bee35daacc97bb2)

You'll find a new template for channels : heating_channel_PID.yaml and a sample config in config.yaml

I mean it could help to stay sync with your repo. 

Regards

**Reply by nliaudat** [reply]
You had BEMF triggers for opening and closing. I have not implemented that cause it's too complicated for lambda to set it up.

**Reply by Karl-opec** [reply]
I use different opening and closing stops, because the motor current curves for opening and closing are quite different. This
gives better control for pop off's of the valve drives.
